slides/08_shiny.qmd

```diff
@@ -0,0 +1,442 @@
+::: {.column width="60%"}
+![](resources/without-modules.png){fig-align="center"}
+:::
+
+::: {.column width="40%"}
+- Difficult to understand what is going on
+- Need longer names / parentheses to differentiate tables / plots / variables
+in different places
+- Hard to test
+:::
+:::
+
+## Modules Example: After
+
+::: columns
+::: {.column width="60%"}
+![](resources/with-modules.png){fig-align="center"}
+:::
+
+::: {.column width="40%"}
+- Can easily see the 4 different parts of the app
+- Naming becomes simpler for vars / tables / plots
+- Can reuse the modules separately in other apps
+- Can test the modules separately
+:::
+:::
+
+## Modules: Considerations similar to Functions
+
+- What should be configurable? 
+  - Think about reuse needs for module
+- How to organize arguments?
+    - Use consistent naming conventions
+    - Keep the right order of the arguments
+- Avoid dependencies between arguments
+  - Use instead parameter object pattern
+- How to write good and maintainable code?
+  - Follow all the clean code rules from the previous chapter!
+  
+# Reactivity
+
+## Reactivity
+
+When building production shiny apps you won't get around learning reactivity.
+
+Some guidance:
+
+- Keep user interface simple → often keeps reactivity simple → often less confusing to user
+- Resolve reactive inputs early on & validate generously
+- Design towards "stringy" reactivity graphs
+- Don't interrupt reactivity
+- Preferably use
+  - `reactive` over `reactiveValues` 
+  - `observeEvent` over `observe`
+
+## Reactivity: Resolve early & validate generously
+
+```{r, echo = TRUE}
+#| code-line-numbers: "3-6,8-13"
+server <- function(input, output, session) {
+  output$plot <- renderPlot({
+    # resolve reactive values
+    xvar <- input$xvar
+    yvar <- input$yvar
+    req(xvar, yvar)
+    
+    # validate – is it likely that the plot will be meaningful?
+    validate(
+      need(xvar %in% names(df), glue("xvar \"{xvar}\" does not exist in data")),
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    my_special_plot(df[[xvar]], df[[yvar]])
+  })
+}
+```
+
+## Reactivity: Debugging with global variables
+
+```{r, echo = TRUE}
+#| code-line-numbers: "16"
+server <- function(input, output, session) {
+  output$plot <- renderPlot({
+    # resolve reactive values
+    xvar <- input$xvar
+    yvar <- input$yvar
+    req(xvar, yvar)
+    
+    # validate – is it likely that the plot will be meaningful?
+    validate(
+      need(xvar %in% names(df), glue("xvar \"{xvar}\" does not exist in data")),
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    .GlobalEnv$xvar <- xvar; .GlobalEnv$yvar <- yvar;  stop()
```

Comment:
shall we use `browser()` or breakpoints in R to enter it? I mean, when debugging outside shiny, I use brower or breakpoints. In shiny this should also work.

Reply:
ok I am adding one more slide on techniques